File: python/3.getaddress.py
# bdkpython 0.26.0
#
# Get new addresses from a wallet.

import bdkpython as bdk
from utilities import check_bdk_version

check_bdk_version(original_major_version=0, original_minor_version=26)


descriptor = bdk.Descriptor(
    "wpkh([c258d2e4/84h/1h/0h]tpubDDYkZojQFQjht8Tm4jsS3iuEmKjTiEGjG6KnuFNKKJb5A6ZUCUZKdvLdSDWofKi4ToRCwb9poe1XdqfUnP4jaJjCB2Zwv11ZLgSbnZSNecE/0/*)",
    bdk.Network.TESTNET,
)
db_config = bdk.DatabaseConfig.MEMORY()

wallet = bdk.Wallet(
             descriptor=descriptor,
             change_descriptor=None,
             network=bdk.Network.TESTNET,
             database_config=db_config,
         )

# print "last unused" receive address
address_info_last_unused = wallet.get_address(bdk.AddressIndex.LAST_UNUSED)
address_last_unused = address_info_last_unused.address
index_last_unused = address_info_last_unused.index
print(f"Last unused P2PKH testnet address: {address_last_unused} at index {index_last_unused}")

# print new receive address
address_info_new = wallet.get_address(bdk.AddressIndex.NEW)
address_new = address_info_new.address
index_new = address_info_new.index
print(f"New P2PKH testnet address: {address_new} at index {index_new}")

File: swift/3.getaddress.swift
#!/usr/bin/swift sh

import BitcoinDevKit  // https://github.com/bitcoindevkit/bdk-swift.git ~0.26.0

class LogProgress : BitcoinDevKit.Progress {
    func update(progress: Float, message: String?) {
        print("Syncing...")
    }
}

do {
    let descriptor = try Descriptor.init(descriptor:"wpkh(tprv8ZgxMBicQKsPf2qfrEygW6fdYseJDDrVnDv26PH5BHdvSuG6ecCbHqLVof9yZcMoM31z9ur3tTYbSnr1WBqbGX97CbXcmp5H6qeMpyvx35B/84h/1h/0h/0/*)", network: .testnet)
    let electrumUrl = "ssl://electrum.blockstream.info:60002"
    let blockchainConfig = BlockchainConfig.electrum(
        config: ElectrumConfig(
            url: electrumUrl,
            socks5: nil,
            retry: 5,
            timeout: nil,
            stopGap: 100,
            validateDomain: true
        )
    )
    let blockchain = try Blockchain(config: blockchainConfig)
    let databaseConfig = DatabaseConfig.memory
    
    let wallet = try Wallet(descriptor: descriptor, changeDescriptor: nil, network: Network.testnet, databaseConfig: databaseConfig)
    let progress = LogProgress()
    try wallet.sync(blockchain: blockchain, progress: progress)

    let lastUnusedAddress = try wallet.getAddress(addressIndex: AddressIndex.lastUnused)
    print("The last unused address for this wallet is \(lastUnusedAddress.address) at index \(lastUnusedAddress.index)")
} catch {
    print(error)
}
